File: python/ntt.py
"""
Number Theoretic Transform over Goldilocks.

In-place radix-2 Cooley-Tukey. The NTT lets us evaluate a degree-(n-1) polynomial
on the n-th roots of unity in O(n log n) instead of O(n^2) Horner evaluations.

Cost on ESP32 is the main bottleneck of any STARK-style prover: a 2^14 NTT in pure
MicroPython is dozens of seconds. In C with Plantard reduction it drops to ~100 ms.
"""

from field import P, inv, mul, add, sub, pow_, primitive_root_of_unity


def _bit_reverse_permute(a):
    """In-place bit-reversal permutation."""
    n = len(a)
    j = 0
    for i in range(1, n):
        bit = n >> 1
        while j & bit:
            j ^= bit
            bit >>= 1
        j |= bit
        if i < j:
            a[i], a[j] = a[j], a[i]


def ntt(a, inverse=False):
    """In-place radix-2 NTT. len(a) must be a power of two. Returns a new list."""
    n = len(a)
    if n == 1:
        return list(a)
    log_n = n.bit_length() - 1
    assert (1 << log_n) == n, "NTT requires power-of-two length"

    a = list(a)
    _bit_reverse_permute(a)

    omega = primitive_root_of_unity(log_n)
    if inverse:
        omega = inv(omega)

    length = 2
    while length <= n:
        # principal length-th root of unity = omega^(n/length)
        w_step = pow_(omega, n // length)
        half = length >> 1
        for i in range(0, n, length):
            w = 1
            for j in range(half):
                u = a[i + j]
                v = (a[i + j + half] * w) % P
                a[i + j] = (u + v) % P
                a[i + j + half] = (u - v) % P
                w = (w * w_step) % P
        length <<= 1

    if inverse:
        n_inv = inv(n)
        a = [(x * n_inv) % P for x in a]
    return a


def coset_ntt(coeffs, shift, target_log_n):
    """
    Evaluate poly `coeffs` on the coset {shift * omega^i : 0 <= i < 2^target_log_n}
    where omega is the primitive (2^target_log_n)-th root of unity.

    Equivalent to: pad coeffs to length 2^target_log_n, multiply c_i <- c_i * shift^i,
    then NTT.
    """
    n = 1 << target_log_n
    assert len(coeffs) <= n, "polynomial larger than target domain"
    padded = list(coeffs) + [0] * (n - len(coeffs))
    cur = 1
    for i in range(n):
        padded[i] = (padded[i] * cur) % P
        cur = (cur * shift) % P
    return ntt(padded, inverse=False)


def coset_intt(values, shift, source_log_n):
    """Inverse of coset_ntt: from evaluations on coset back to coefficients."""
    n = 1 << source_log_n
    assert len(values) == n
    b = ntt(values, inverse=True)
    shift_inv = inv(shift)
    cur = 1
    for i in range(n):
        b[i] = (b[i] * cur) % P
        cur = (cur * shift_inv) % P
    return b

File: c/ntt.c
#include "ntt.h"

static void bit_reverse_permute(felt *a, uint32_t n) {
    uint32_t j = 0;
    for (uint32_t i = 1; i < n; ++i) {
        uint32_t bit = n >> 1;
        for (; j & bit; bit >>= 1) j ^= bit;
        j |= bit;
        if (i < j) { felt t = a[i]; a[i] = a[j]; a[j] = t; }
    }
}

void ntt_inplace(felt *a, uint8_t log_n, int inverse) {
    uint32_t n = 1u << log_n;
    if (n == 1) return;

    bit_reverse_permute(a, n);

    felt omega = f_root_of_unity(log_n);
    if (inverse) omega = f_inv(omega);

    for (uint32_t length = 2; length <= n; length <<= 1) {
        /* principal length-th root = omega^(n/length) */
        felt w_step = f_pow(omega, (uint64_t)(n / length));
        uint32_t half = length >> 1;
        for (uint32_t i = 0; i < n; i += length) {
            felt w = 1;
            for (uint32_t k = 0; k < half; ++k) {
                felt u = a[i + k];
                felt v = f_mul(a[i + k + half], w);
                a[i + k]        = f_add(u, v);
                a[i + k + half] = f_sub(u, v);
                w = f_mul(w, w_step);
            }
        }
    }

    if (inverse) {
        felt n_inv = f_inv(f_from_u64((uint64_t)n));
        for (uint32_t i = 0; i < n; ++i) a[i] = f_mul(a[i], n_inv);
    }
}

void coset_eval(const felt *coeffs, uint8_t log_n_src,
                felt shift, uint8_t log_N, felt *out) {
    uint32_t n = 1u << log_N;
    uint32_t src = 1u << log_n_src;
    /* copy + zero-pad */
    felt cur = 1;
    for (uint32_t i = 0; i < n; ++i) {
        felt c = (i < src) ? coeffs[i] : 0;
        out[i] = f_mul(c, cur);   /* c_i * shift^i */
        cur = f_mul(cur, shift);
    }
    ntt_inplace(out, log_N, 0);
}

void coset_interp(const felt *evals, felt shift, uint8_t log_N, felt *out) {
    uint32_t n = 1u << log_N;
    for (uint32_t i = 0; i < n; ++i) out[i] = evals[i];
    ntt_inplace(out, log_N, 1);
    felt shift_inv = f_inv(shift);
    felt cur = 1;
    for (uint32_t i = 0; i < n; ++i) {
        out[i] = f_mul(out[i], cur);
        cur = f_mul(cur, shift_inv);
    }
}
